xds/src/main/java/io/grpc/xds/Filter.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Any;
+import io.grpc.ClientInterceptor;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.ServerInterceptor;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * Defines the parsing functionality of an HTTP filter. A Filter may optionally implement either
```

Comment:
Such an interface looks weird. It seems to be really just a Parser, as it is documented as "Defines the parsing functionality of an HTTP filter...". Then you probably want some interface to represent `Filter` objects (generics may help).

Reply:
This is aligned with  [C-core](https://github.com/grpc/grpc/blob/master/src/core/ext/xds/xds_http_filters.h#L77) and [Go](https://github.com/grpc/grpc-go/blob/master/xds/internal/httpfilter/httpfilter.go#L46) APIs. I don't want to deviate too much from APIs of other languages.

Reply:
The API spec in the [design doc](https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md#http-filter-registry)

```
Each filter implementation will provide the following:

Method(s) for validating the config protos when parsing an xDS response.
An indication of whether it is supported on the gRPC client and/or gRPC server.
Methods to generate and configure the appropriate filters or interceptors in gRPC to perform the necessary functionality on the data plane.
```